### PR TITLE
Refactor blueprints_updater: ID generation, caching, and safety

### PR DESCRIPTION
**Pull request by luuquangvu**
## Summary by Sourcery

Improve blueprint update robustness, safety, and configurability while adding deterministic IDs, better caching, and multi‑entry support across the integration.

New Features:
- Add deterministic and legacy‑compatible unique ID generation for blueprint update entities and service handlers.
- Introduce centralized helpers for normalized update interval and backup count configuration, with clamping and coercion.
- Support multiple config entries/coordinators in services and update entities, including per‑entry bulk update handling.

Bug Fixes:
- Prevent unsafe filesystem writes and restores by tightening path validation and using real paths for security checks.
- Handle corrupted or invalid persisted metadata and translation data safely, avoiding crashes and pruning stale entries.
- Ensure background refresh workers and HTTP retries handle cancellations and redirect security correctly, preventing hangs or unsafe redirects.
- Fix cases where unsafe or empty remote content could leave stale remote metadata or incorrect update status in coordinator state.
- Resolve entity/ID mismatches by migrating legacy unique IDs and correctly mapping entities to their owning config entries.

Enhancements:
- Refine backup rotation and restore logic to be more robust, reversible, and tolerant of filesystem errors.
- Improve blueprint validation, error reporting, and state tracking (including invalid remote hashes and structured last_error values).
- Strengthen URL and IP safety checks, including multicast/reserved ranges and forum content parsing.
- Add shared translation caching and clearing hooks for both the coordinator and setup layer to reduce repeated translation lookups.
- Rework bulk update and single‑install flows to fetch missing content on demand, continue past failures, and keep coordinator data in sync.
- Enhance retry_async with stricter argument validation, explicit exception lists, better logging context, and cancellation propagation.
- Tighten blueprint scanning and source_url injection to handle more YAML shapes and avoid false positives on non‑blueprint keys.

Build:
- Enable pydocstyle checks via Ruff and configure the project to use the Google docstring convention.

Documentation:
- Adopt Google‑style docstring linting and update docstrings across the integration for clearer parameter and return value descriptions.

Tests:
- Expand and adjust test coverage for coordinator behavior, backup/restore, HTTP pacing and retries, unsafe URL/path handling, and metadata pruning.
- Add tests for config flow defaults and clamping, translation cache clearing, multi‑coordinator service selection, and bulk update flows.
- Strengthen update entity tests around release notes, usage calculation errors, legacy ID migration, and install error handling.
- Extend utils tests to cover retry_async edge cases and new configuration normalization helpers.

## Summary by CodeRabbit

* **New Features**
  * Deterministic entity IDs with automatic migration; per-coordinator background update processing and on-demand fetching.

* **Bug Fixes**
  * Translation caching with automatic invalidation on core config changes.
  * Restore/update target correct coordinator; safer installs, backup rotation, redirect/safety checks, and clearer empty/unsafe-content errors.

* **Refactor**
  * Safer option coercion/clamping, tightened retry behavior, and stricter validation/guards.

* **Tests**
  * Expanded coverage: flows, coordinator lifecycle, fetch/etag/backup logic, retries, caching, migrations, and restore/update behavior.

### CODE REVIEW

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

File: custom_components/blueprints_updater/__init__.py [hunk 1]
```diff
@@ -143,39 +156,44 @@ async def _translate(key: str, category: str = "exceptions", **kwargs: str) -> s
...
+            raise ServiceValidationError(
+                translation_domain=DOMAIN,
+                translation_key="not_found",
+            )
+
         target_path = None
         for path, info in active_coordinator.data.items():
-            expected_id = f"blueprint_{hashlib.sha256(info['rel_path'].encode()).hexdigest()}"
+            expected_id = BlueprintUpdateCoordinator.generate_unique_id(
+                active_coordinator.config_entry.entry_id, info["rel_path"]
```
Comment (SourceryAI) [inline custom_components/blueprints_updater/__init__.py:210]: **issue:** Service restore uses only the new unique_id format and will not work for legacy entities until after migration runs.

In `async_restore_blueprint_handler`, this now only matches on `generate_unique_id(entry_id, rel_path)`. Existing entities that still use the legacy `rel_path`-only unique_id won’t be found until the migration in `update.async_setup_entry` completes, so `invalid_blueprint` can be raised even though the entity exists. Consider mirroring `async_update_entities` by also computing `generate_legacy_unique_id(rel_path)` and matching against both formats.

✅ Addressed in 54ab1ded91941641891fa83d651696219f5db08b: `async_restore_blueprint_handler` now checks both `generate_unique_id(entry_id, rel_path)` and `generate_legacy_unique_id(rel_path)`, so legacy entities are supported during restore.

File: tests/test_init.py [hunk 2]
```diff
@@ -12,13 +13,18 @@
     async_unload_entry,
 )
 from custom_components.blueprints_updater.const import DOMAIN
+from custom_components.blueprints_updater.coordinator import BlueprintUpdateCoordinator
```
Comment (SourceryAI) [inline tests/test_init.py:16]: **suggestion (testing):** Add direct unit tests for `generate_unique_id` vs `generate_legacy_unique_id`

Current tests exercise these IDs indirectly via the service handler and migration paths, but don’t pin down their semantics. Please add a focused unit test (e.g. in `test_coordinator.py`) that:

- Verifies `generate_unique_id(entry_id, rel_path)` is deterministic and varies with `entry_id` when `rel_path` is fixed.
- Verifies `generate_unique_id(entry_id, rel_path)` is different from `generate_legacy_unique_id(rel_path)` for the same path.

This will surface unintended format changes that could impact migrations and service lookups.

```suggestion
from custom_components.blueprints_updater.const import DOMAIN
from custom_components.blueprints_updater.coordinator import (
    BlueprintUpdateCoordinator,
    generate_legacy_unique_id,
    generate_unique_id,
)


def test_generate_unique_id_deterministic_and_entry_specific() -> None:
    """generate_unique_id should be deterministic and depend on entry_id."""
    relative_path = "blueprints/automation/test_blueprint.yaml"
    entry_id_1 = "entry-id-1"
    entry_id_2 = "entry-id-2"

    unique_id_1_first = generate_unique_id(entry_id_1, relative_path)
    unique_id_1_second = generate_unique_id(entry_id_1, relative_path)
    unique_id_2 = generate_unique_id(entry_id_2, relative_path)

    # Deterministic for the same inputs
    assert unique_id_1_first == unique_id_1_second

    # Varies with entry_id when relative_path is fixed
    assert unique_id_1_first != unique_id_2


def test_generate_unique_id_differs_from_legacy_unique_id() -> None:
    """New unique_id format must differ from legacy format for same path."""
    relative_path = "blueprints/automation/test_blueprint.yaml"
    entry_id = "entry-id-1"

    new_unique_id = generate_unique_id(entry_id, relative_path)
    legacy_unique_id = generate_legacy_unique_id(relative_path)

    # Guard against unintended convergence of formats that would break migrations
    assert new_unique_id != legacy_unique_id
```

**Review by SourceryAI — COMMENTED** [review]
Hey - I've found 2 issues

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location path="custom_components/blueprints_updater/__init__.py" line_range="194-195" />
<code_context>
         target_path = None
         for path, info in active_coordinator.data.items():
-            expected_id = f"blueprint_{hashlib.sha256(info['rel_path'].encode()).hexdigest()}"
+            expected_id = BlueprintUpdateCoordinator.generate_unique_id(
+                active_coordinator.config_entry.entry_id, info["rel_path"]
+            )
             if expected_id == entity_entry.unique_id:
</code_context>
<issue_to_address>
**issue:** Service restore uses only the new unique_id format and will not work for legacy entities until after migration runs.

In `async_restore_blueprint_handler`, this now only matches on `generate_unique_id(entry_id, rel_path)`. Existing entities that still use the legacy `rel_path`-only unique_id won’t be found until the migration in `update.async_setup_entry` completes, so `invalid_blueprint` can be raised even though the entity exists. Consider mirroring `async_update_entities` by also computing `generate_legacy_unique_id(rel_path)` and matching against both formats.
</issue_to_address>

### Comment 2
<location path="tests/test_init.py" line_range="15-16" />
<code_context>
     async_unload_entry,
 )
 from custom_components.blueprints_updater.const import DOMAIN
+from custom_components.blueprints_updater.coordinator import BlueprintUpdateCoordinator


</code_context>
<issue_to_address>
**suggestion (testing):** Add direct unit tests for `generate_unique_id` vs `generate_legacy_unique_id`

Current tests exercise these IDs indirectly via the service handler and migration paths, but don’t pin down their semantics. Please add a focused unit test (e.g. in `test_coordinator.py`) that:

- Verifies `generate_unique_id(entry_id, rel_path)` is deterministic and varies with `entry_id` when `rel_path` is fixed.
- Verifies `generate_unique_id(entry_id, rel_path)` is different from `generate_legacy_unique_id(rel_path)` for the same path.

This will surface unintended format changes that could impact migrations and service lookups.

```suggestion
from custom_components.blueprints_updater.const import DOMAIN
from custom_components.blueprints_updater.coordinator import (
    BlueprintUpdateCoordinator,
    generate_legacy_unique_id,
    generate_unique_id,
)


def test_generate_unique_id_deterministic_and_entry_specific() -> None:
    """generate_unique_id should be deterministic and depend on entry_id."""
    relative_path = "blueprints/automation/test_blueprint.yaml"
    entry_id_1 = "entry-id-1"
    entry_id_2 = "entry-id-2"

    unique_id_1_first = generate_unique_id(entry_id_1, relative_path)
    unique_id_1_second = generate_unique_id(entry_id_1, relative_path)
    unique_id_2 = generate_unique_id(entry_id_2, relative_path)

    # Deterministic for the same inputs
    assert unique_id_1_first == unique_id_1_second

    # Varies with entry_id when relative_path is fixed
    assert unique_id_1_first != unique_id_2


def test_generate_unique_id_differs_from_legacy_unique_id() -> None:
    """New unique_id format must differ from legacy format for same path."""
    relative_path = "blueprints/automation/test_blueprint.yaml"
    entry_id = "entry-id-1"

    new_unique_id = generate_unique_id(entry_id, relative_path)
    legacy_unique_id = generate_legacy_unique_id(relative_path)

    # Guard against unintended convergence of formats that would break migrations
    assert new_unique_id != legacy_unique_id
```
</issue_to_address>
~~~

</details>

***

Hi @luuquangvu! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai summary

**Comment by luuquangvu** [comment]
@sourcery-ai summary

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@sourcery-ai summary

**Comment by luuquangvu** [comment]
@sourcery-ai guide

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai resolve

**Comment by luuquangvu** [comment]
@coderabbitai resolve

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@sourcery-ai resolve

**Comment by luuquangvu** [comment]
@coderabbitai resolve

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
Close #39
